lib/handler/retry-handler.js

```diff
@@ -59,292 +70,227 @@ class RetryHandler {
+    this.#retryCount += 1
+
+    if (this.#aborted) {
+      abort(this.#reason)
     } else {
       this.abort = abort
     }
   }
 
-  onBodySent (chunk) {
-    if (this.handler.onBodySent) return this.handler.onBodySent(chunk)
-  }
-
-  static [kRetryHandlerDefaultRetry] (err, { state, opts }, cb) {
-    const { statusCode, code, headers } = err
-    const { method, retryOptions } = opts
-    const {
-      maxRetries,
-      minTimeout,
-      maxTimeout,
-      timeoutFactor,
-      statusCodes,
-      errorCodes,
-      methods
-    } = retryOptions
-    const { counter } = state
-
-    // Any code that is not a Undici's originated and allowed to retry
-    if (
-      code &&
-      code !== 'UND_ERR_REQ_RETRY' &&
-      !errorCodes.includes(code)
-    ) {
-      cb(err)
-      return
-    }
-
-    // If a set of method are provided and the current method is not in the list
-    if (Array.isArray(methods) && !methods.includes(method)) {
-      cb(err)
-      return
-    }
-
-    // If a set of status code are provided and the current status code is not in the list
-    if (
-      statusCode != null &&
-      Array.isArray(statusCodes) &&
-      !statusCodes.includes(statusCode)
-    ) {
-      cb(err)
-      return
-    }
-
-    // If we reached the max number of retries
-    if (counter > maxRetries) {
-      cb(err)
-      return
-    }
-
-    let retryAfterHeader = headers?.['retry-after']
-    if (retryAfterHeader) {
-      retryAfterHeader = Number(retryAfterHeader)
-      retryAfterHeader = Number.isNaN(retryAfterHeader)
-        ? calculateRetryAfterHeader(retryAfterHeader)
-        : retryAfterHeader * 1e3 // Retry-After is in seconds
-    }
-
-    const retryTimeout =
-      retryAfterHeader > 0
-        ? Math.min(retryAfterHeader, maxTimeout)
-        : Math.min(minTimeout * timeoutFactor ** (counter - 1), maxTimeout)
-
-    setTimeout(() => cb(null), retryTimeout)
-  }
-
   onHeaders (statusCode, rawHeaders, resume, statusMessage) {
     const headers = parseHeaders(rawHeaders)
 
-    this.retryCount += 1
-
     if (statusCode >= 300) {
-      if (this.retryOpts.statusCodes.includes(statusCode) === false) {
-        return this.handler.onHeaders(
-          statusCode,
-          rawHeaders,
-          resume,
-          statusMessage
-        )
-      } else {
-        this.abort(
-          new RequestRetryError('Request failed', statusCode, {
-            headers,
-            count: this.retryCount
-          })
-        )
-        return false
-      }
-    }
-
-    // Checkpoint for resume from where we left it
-    if (this.resume != null) {
-      this.resume = null
-
-      if (statusCode !== 206) {
-        return true
+      if (this.#retryOpts.statusCodes.includes(statusCode)) {
+        // TODO (fix): Remove this path and delegate to different interceptor which will read the error body?
+        const message = `Response status code ${statusCode}${statusMessage ? `: ${statusMessage}` : ''}`
```

Comment:
Curious about this one, what did you have in mind @ronag?

Reply:
Should propagate similar to throwOnError